npep/npep-122.md

```diff
@@ -0,0 +1,181 @@
+    facilitate network multi-tenancy and strict network level isolation between multiple teams and tenants sharing 
+    a cluster via use of namespaces or groupings of namespaces per tenant.
+    
+    Example: I would like to define two tenants in my cluster, one composed of the pods in foo-ns-1 and foo-ns-2 
+    and the other with pods in bar-ns-1, where inter-tenant traffic is denied.
+
+From the User Stories:
+
+    Story 4: Create and Isolate multiple tenants in a cluster
+    
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default. 
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+Elsewhere:
+
+    AdminNetworkPolicy Pass rules allows an admin to delegate security posture for certain traffic to the Namespace 
+    owners by overriding any lower precedence Allow or Deny rules. For example, intra-tenant traffic management can be 
+    delegated to tenant admins explicitly with the use of Pass rules.
+    
+So really, the only solidly-agreed-upon use case is that either you should be able to create enforced isolation 
+between particular sets of namespaces.
+
+### What we currently have
+
+AdminNetworkPolicy has the “SameLabels” and “NotSameLabels” fields to support the use cases involving tenancy. For example:
+
+**User story**
+
+Traffic should be disallowed by default between user namespaces owned by different users.
+
+**Policy**
+```
+kind: BaselineAdminNetworkPolicy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based-tenancy
+spec:
+  subject:
+    namespaces:
+      matchExpressions:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces:
+            notSameLabels:
+              - user
+```
+
+**Meaning**
+
+In namespaces that have a “user” label, by default, deny ingress from namespaces that have a different “user” label.
+
+There are two problems with this implementation of tenancy, relative to the user stories that it exists to support
+
+First, it never explicitly defines what a tenant is. Although the object correctly implements the desired tenancy 
+policy, it doesn’t really make it obvious that that’s what it’s doing.
+
+Second, the syntax is very general purpose / powerful. ANP has subjects and peers, which are different, 
+and currently Tenancy is defined on the peers side. Tenancy by itself has the same subject and peer, 
+at least for the existing use cases, and having separate selectors for subject and peer allows for more
+configurations than needed.
+
+## API
+
+We suggest to define a separate API for tenancy, as in the following example:
+
+```
+kind: AdminNetworkPolicyTenancy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based
+spec:
+  namespaceSelector:
```

Comment:
Maybe this would be `tenantSubject` or something? Although I don't really want to hash about API design just yet 🤣